### PR TITLE
Finishing SLAP data-flow (adds action prediction module)

### PR DESCRIPTION
**Pull request by zephirefaith**
## Motivation and Context
2nd part of ongoing series to port SLAP to Stretch infrastructure. The aim of this PR is to finish the inference data-flow pipeline. We are able to:
1. Provide perception from stretch to interaction prediction module
2. Execute end-effector pose, quat goal coming from an H5 on robot
This PR then adds:
- [x] Pipe output of IPM + perception into action prediction modue
~- [ ] Replace end-effector pose coming from H5 files with output of SLAP~

## How Has This Been Tested

Testing on the robot.

## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

# Checklist:

- [ ] I have made changes to existing documentation where needed.
- [ ] I have added tests that show that the PR is functional.
- [ ] I ran `tests/hw_manual_test.py` on hardware and observed that robot behavior looks normal.

### CODE REVIEW

File: projects/slap_manipulation/scripts/conf/test.yaml [hunk 1]
```diff
@@ -1,5 +1,8 @@
 dry_run: True
 ipm_weights: /home/priparashar/classify-multi-task_0035_000388.pth
```
Comment (cpaxton) [inline projects/slap_manipulation/scripts/conf/test.yaml:2]: we shouldn't commit paths like this

File: projects/slap_manipulation/scripts/conf/test.yaml [hunk 2]
```diff
@@ -1,5 +1,8 @@
...
+num_keypoints: 3
```
Comment (cpaxton) [inline projects/slap_manipulation/scripts/conf/test.yaml:3]: should we add descriptions for these parameters?

File: projects/slap_manipulation/scripts/slap_inference.py [hunk 3]
```diff
@@ -15,27 +20,61 @@
 STRETCH_GRIPPER_MAX = 0.6
 
 
-def create_apm_input(raw_data, p_i):
+def get_proprio(raw_data, time=0.0):
```
Comment (cpaxton) [inline projects/slap_manipulation/scripts/slap_inference.py:23]: i still don't know if i like this name? proprio was just the term they used in per-act

File: projects/slap_manipulation/scripts/slap_inference.py [hunk 4]
```diff
@@ -15,27 +20,61 @@
...
+    proprio = raw_data["gripper_state"].astype(np.float64)
+    if proprio < 0.8 * STRETCH_GRIPPER_MAX:
+        proprio = np.array([1.0, proprio, time])
+    else:
+        proprio = np.array([0.0, proprio, time])
+
+    return proprio
+
+
+def create_apm_input(
```
Comment (cpaxton) [inline projects/slap_manipulation/scripts/slap_inference.py:39]: should we rename create_apm to something a bit clearer? create_action_prediction_input?

File: projects/slap_manipulation/scripts/slap_inference.py [hunk 5]
```diff
@@ -15,27 +20,61 @@
...
+    )
+    if not status:
+        raise RuntimeError(
+            "Interaction Prediction Module predicted an interaction point with no tractable local problem around it"
+        )
+    proprio = get_proprio(raw_data, time=time)
+    return (cropped_feat, cropped_xyz, proprio)
 
 
 def create_ipm_input(
```
Comment (cpaxton) [inline projects/slap_manipulation/scripts/slap_inference.py:61]: same here - avoid using acronyms too heavily imo, makes it less readable

File: projects/slap_manipulation/scripts/slap_inference.py [hunk 6]
```diff
@@ -83,7 +122,7 @@ def main(cfg):
     apm_model = ActionPredictionModule(dry_run=cfg.dry_run)
     # load model-weights
     ipm_model.load_state_dict(torch.load(cfg.ipm_weights))
```
Comment (cpaxton) [inline projects/slap_manipulation/scripts/slap_inference.py:124]: interaction_predictor and action_predictor would be clearer again

File: projects/slap_manipulation/src/slap_manipulation/dataloaders/robot_loader.py [hunk 7]
```diff
@@ -464,6 +464,7 @@ def get_datum(self, trial, keypoint_idx, verbose=False):
                 xyzs.append(v_xyz)
 
         # randomly dropout 1/3rd of the point-clouds
+        # TODO: update this to dropout each frame with 0.33 probability
```
Comment (cpaxton) [inline projects/slap_manipulation/src/slap_manipulation/dataloaders/robot_loader.py:467]: add issue? or just implement this

File: projects/slap_manipulation/src/slap_manipulation/utils/input_preprocessing.py [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+
+
+def get_local_action_prediction_problem(
+    cfg,
+    feat: np.ndarray,
+    xyz: np.ndarray,
+    p_i: np.ndarray,
+) -> Tuple[np.ndarray, np.ndarray, bool]:
+    """
+    Create local problem around predicted interaction point for regression training
```
Comment (cpaxton) [inline projects/slap_manipulation/src/slap_manipulation/utils/input_preprocessing.py:26]: what is a "local problem"

its a mean-centered point cloud around p_i right

Comment (cpaxton) [reply projects/slap_manipulation/src/slap_manipulation/utils/input_preprocessing.py:26]: to be clear i know the answer i just think this comment isnt super helpful

**Review by cpaxton — COMMENTED** [review]
minor doc fixes and a couple questions

File: projects/slap_manipulation/scripts/conf/test.yaml (hunk 1 above)
Comment (zephirefaith) [reply projects/slap_manipulation/scripts/conf/test.yaml:2]: Good callout! Totally missed this.

File: projects/slap_manipulation/scripts/slap_inference.py (hunk 3 above)
Comment (zephirefaith) [reply projects/slap_manipulation/scripts/slap_inference.py:23]: It is also a weird vector: it has a bool gripper-action, a float gripper-width and the time of current keyframe wrt entire task.

Maybe split it up into `[robot/gripper]_state` and `task_state` (just the time)?

**Review by exhaustin — APPROVED** [review]
Approving since this PR is pure project development and we don't want the bar for `projects/` to be too high because it's a place for experimental code.

However, it would be nice to add some minimal documentions in terms of installation and entry point for running experiments.

File: projects/slap_manipulation/scripts/slap_inference.py (hunk 3 above)
Comment (cpaxton) [reply projects/slap_manipulation/scripts/slap_inference.py:23]: thats fine with me

**Review by cpaxton — APPROVED** [review]
lgtm

**Pull request merged**